File: conventional-commits-common/src/main/java/com/smartling/cc4j/semantic/release/common/GitConventionalVersioning.java
package com.smartling.cc4j.semantic.release.common;

import com.smartling.cc4j.semantic.release.common.scm.ScmApiException;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.revwalk.RevCommit;

import java.io.IOException;
import java.util.Objects;

public class GitConventionalVersioning implements ConventionalVersioning
{
    private final Repository repository;

    public GitConventionalVersioning(Repository repository)
    {
        Objects.requireNonNull(repository, "repository required");
        this.repository = repository;
    }

    @Override
    public LogHandler logHandler()
    {
        return new LogHandler(repository);
    }

    private SemanticVersionChangeResolver semanticVersionChangeResolver()
    {
        return new SimpleSemanticVersionChangeResolver();
    }

    @Override
    public SemanticVersionChange getNextVersionChangeType() throws ScmApiException, IOException
    {
        try
        {
            final Iterable<RevCommit> commits = logHandler().getCommitsSinceLastTag();
            final SemanticVersionChangeResolver resolver = semanticVersionChangeResolver();

            return resolver.resolveChange(commits);
        }
        catch (final GitAPIException e)
        {
            throw new ScmApiException("Git operation failed", e);
        }
    }

    @Override
    public SemanticVersion getNextVersion(SemanticVersion currentVersion) throws IOException, ScmApiException
    {
        final SemanticVersionChange change = this.getNextVersionChangeType();
        return currentVersion.nextVersion(change);
    }
}

File: conventional-commits-common/src/main/java/com/smartling/cc4j/semantic/release/common/GitConventionalVersioning.kt
package com.smartling.cc4j.semantic.release.common

import kotlin.Throws
import com.smartling.cc4j.semantic.release.common.scm.ScmApiException
import java.io.IOException
import org.eclipse.jgit.api.errors.GitAPIException
import com.smartling.cc4j.semantic.release.common.SemanticVersion
import org.eclipse.jgit.lib.Repository
import java.util.*

class GitConventionalVersioning(private val repository: Repository) : ConventionalVersioning {
    private val semanticVersionChangeResolver = SimpleSemanticVersionChangeResolver()

    override fun logHandler(): LogHandler {
        return LogHandler(this.repository)
    }

    @Throws(ScmApiException::class, IOException::class)
    override fun getNextVersionChangeType(): SemanticVersionChange {
        return try {
            val commits = logHandler().getCommitsSinceLastTag()
            this.semanticVersionChangeResolver.resolveChange(commits)
        } catch (e: GitAPIException) {
            throw ScmApiException("Git operation failed", e)
        }
    }

    @Throws(IOException::class, ScmApiException::class)
    override fun getNextVersion(currentVersion: SemanticVersion): SemanticVersion {
        val change = this.getNextVersionChangeType()
        return currentVersion.nextVersion(change)
    }
}
